### PR TITLE
Job use deploy agent

### PR DESCRIPTION
**Pull request by arthurcgc**

### CODE REVIEW

File: job/job.go [hunk 1]
```diff
@@ -128,6 +208,18 @@ func (*jobService) CreateJob(ctx context.Context, job *jobTypes.Job, user *authT
...
 	}
 
+	// use deploy-agent to push the image to tsuru's registry
+	newImageDst, err := builderDeploy(ctx, job, builder.BuildOpts{
+		ImageID: job.Spec.Container.Image,
+	})
+	// we don't want to fail the job creation if the image push fails
+	if err == nil {
+		// deploy the job using the new pushed image
+		job.Spec.Container.Image = newImageDst
```
Comment (wpjunior) [inline job/job.go:218]: @wpjunior store this outside of Spec, like
job.Status.Image or job.Status.TsuruImage or job.Status.PushedImage

File: types/deploy/options/deploy_options.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package options
```
Comment (wpjunior) [inline types/deploy/options/deploy_options.go:1]: why not move this file to types/provision/deploy_options.go ?

File: job/job.go (hunk 1 above)
Comment (arthurcgc) [reply job/job.go:218]: I forgot to open the PR as draft, still making changes

File: types/deploy/options/deploy_options.go (hunk 2 above)
Comment (arthurcgc) [reply types/deploy/options/deploy_options.go:1]: I forgot to open the PR as draft, still making changes

File: job/job.go (hunk 1 above)
Comment (arthurcgc) [reply job/job.go:218]: Also, it makes more sense to store the original image outside of spec and keep the deploy-agent image inside spec

Comment (wpjunior) [reply job/job.go:218]: kubernetes stores human data on Spec, them processed data on Status, I enjoy this convention

Comment (arthurcgc) [reply job/job.go:218]: Actually, we pass JobSpec to the provisioner and honor it, k8s takes the JobSpec specified and creates the Cronjob object, it's certainly not usual nor is it necessary to have a divergent Spec and Status, there's no need for it. If we want to "save" the original image source it's fine, but we should keep the image being used in JobSpec

File: api/deploy.go [hunk 3]
```diff
@@ -18,6 +18,7 @@ import (
 	"github.com/tsuru/tsuru/event"
 	tsuruIo "github.com/tsuru/tsuru/io"
 	"github.com/tsuru/tsuru/permission"
+	deployOpts "github.com/tsuru/tsuru/types/provision"
```
Comment (wpjunior) [inline api/deploy.go:21]: @arthurcgc could you rename the symbol to provisionTypes?

File: job/job.go [hunk 4]
```diff
@@ -101,6 +106,84 @@ func (*jobService) RemoveJobProv(ctx context.Context, job *jobTypes.Job) error {
...
+func buildWithDeployAgent(ctx context.Context, job *jobTypes.Job) error {
+	// use deploy-agent to push the image to tsuru's registry
+	newImageDst, err := builderDeploy(ctx, job, builder.BuildOpts{
+		ImageID: job.Spec.Container.Image,
+	})
+	// we don't want to fail the job creation if the image push fails, yet
+	if err == nil && newImageDst != "" {
+		// deploy the job using the new pushed image
+		job.OriginalImageURL = job.Spec.Container.Image
+		job.Spec.Container.Image = newImageDst
```
Comment (wpjunior) [inline job/job.go:181]: this behavior will fail terraform providers, we must store `MirrorImageURL` and maintain intact the image provided by the user. @arthurcgc

File: types/job/job.go [hunk 5]
```diff
@@ -21,14 +21,15 @@ var _ provisionTypes.ResourceGetter = &Job{}
...
-	Description string            `json:"description"`
+	Name             string            `json:"name"`
+	Teams            []string          `json:"teams"`
+	TeamOwner        string            `json:"teamOwner"`
+	Owner            string            `json:"owner"`
+	Plan             appTypes.Plan     `json:"plan"`
+	Metadata         appTypes.Metadata `json:"metadata"`
+	Pool             string            `json:"pool"`
+	Description      string            `json:"description"`
+	OriginalImageURL string            `json:"originalImageURL"`
```
Comment (wpjunior) [inline types/job/job.go:32]: @arthurcgc please drop OriginalImageURL and create MirrorImageURL

File: types/provision/deploy_options.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package provision
```
Comment (wpjunior) [inline types/provision/deploy_options.go:1]: @arthurcgc missing LICENCSE Header

File: types/job/job.go [hunk 7]
```diff
@@ -63,8 +63,9 @@ func (job *Job) GetPool() string {
 }
 
 type ContainerInfo struct {
-	Image   string   `json:"image"`
-	Command []string `json:"command"`
+	InternalRegistryImage string   `json:"internalRegistryImage"`
```
Comment (wpjunior) [inline types/job/job.go:66]: rename a field causes a interoperability problem @arthurcgc

**Review by wpjunior — APPROVED** [review]

**Pull request merged**